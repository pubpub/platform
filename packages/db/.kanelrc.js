// @ts-check
const { makeKyselyHook } = require("kanel-kysely");
const { generateZodSchemas } = require("kanel-zod");
const {
	kanelKyselyZodCompatibilityPreRenderHook,
} = require("./src/kanel/kanel-kysely-zod-compatibility-hook.cjs");
const {
	kanelDatabaseDefaultExportFixPreRenderHook,
} = require("./src/kanel/kanel-database-default-export-fix-hook.cjs");
const { cleanupEnumComments } = require("./src/kanel/kanel-cleanup-enum-comments.cjs");
const { escapeIdentifier } = require("kanel");
const { resolveType } = require("kanel");
const { escapeString } = require("kanel");
const { recase } = require("@kristiandupont/recase");

const toPascalCase = recase(null, "pascal");

/** @type {import('kanel').Config} */
module.exports = {
	connection: process.env["DATABASE_URL"],
	schemas: ["public"],

	preDeleteOutputFolder: false,
	enumStyle: "enum",
	generateIdentifierType: (column, details, config) => {
		const name = escapeIdentifier(toPascalCase(details.name) + toPascalCase(column.name));
		const innerType = resolveType(column, details, {
			...config,
			// Explicitly disable identifier resolution so we get the actual inner type here
			generateIdentifierType: undefined,
		});
		const imports = [];

		let type = innerType;
		if (typeof innerType === "object") {
			// Handle non-primitives
			type = innerType.name;
			imports.push(...innerType.typeImports);
		}

		let typeDefinition = [`${type} & { __brand: '${escapeString(name)}' }`];

		/* this is the custom part */
		if (column.type.kind === "enum") {
			typeDefinition = [`${type}`];
		}

		return {
			declarationType: "typeDeclaration",
			name,
			exportAs: "named",
			typeDefinition,
			typeImports: imports,
			comment: [`Identifier type for ${details.schemaName}.${details.name}`],
		};
	},
	preRenderHooks: [
		makeKyselyHook(),
		generateZodSchemas,
		kanelKyselyZodCompatibilityPreRenderHook,
		kanelDatabaseDefaultExportFixPreRenderHook,
		cleanupEnumComments,
	],
	outputPath: "./src",

	customTypeMap: {
		"pg_catalog.tsvector": "string",
		"pg_catalog.bpchar": "string",
	},
};
