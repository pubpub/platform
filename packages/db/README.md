# db

Store the types from our database in a single place.

They are generated using `kanel`, `kanel-kysely`, and `kanel-zod` from our postgres database.

## Generating the types

Make sure the database is running on `localhost:54322` and the user is `postgres`.

The default way of doing this is by following the instruction in `core/README.md`.

Or, if you don't have it run there, create an `.env.local` file with the following content:

```bash
DATABASE_URL="your postgres url"
```

To generate just the types, run

```bash
pnpm --filter db make-kysely-types
```

If you want to regenerate the types after updating `core/prisma/schema.prisma`, run

```bash
pnpm --filter db migrate-dev
```

## Adding a table

If you have added a new table, make sure to add

```ts
export * from "./tableName";
```

to `src/public.ts`, otherwise you will not be able to import it.

## Special hooks

Because `kanel` is a bit messy, we have a few special hooks to make it easier to work with.

See `src/kanel` for more details.

## Adding extra types you want to use

You can add/export extra types in `src/types`. This can be useful if you want to augment the default types that `kanel` generates, for instance for
JSON columns.

Ex.

By default, `kanel` sets the type of a JSON column to `unknown`.

You can change this by adding an `/// @type()` annotation in the `core/schema/schema.prisma` file, which will get translated into a comment on that column in Postgres. That comment will get converted into a type annotation with `kanel`.

See for instance this comment in `core/schema/schema.prisma`:

```prisma
  constraints      Json? /// @type(ApiAccessPermissionConstraints, '../types', true, false, true)
```

This will turn the `contraints` column of `ApiAccessPermission` into `ApiAccessPermissionConstraints`, imported from `../types`.

See `core/kysely/README.md` for slightly more info on this.

TODO: all our database notes should probably be centralized somewhere.

## History Tables

We have a number of tables that exist just to record the history of a table, along with the entity that made the change. They serve both as a way to audit changes and as a way to reconstruct the value of an entity (usually a Pub) at a point in time.

### How to create one

1. Run `pnpm db:generate-history-table`
2. Enter the name of the original table (e.g. `pubs`)
3. Say yes every time
4. Run `pnpm db:migrate-dev` (probably twice for good measure)
5. Tadah! You should have a new table called `pubs_history`

This will also add a new column to the original table called `lastModifiedBy` of type `LastModifiedBy`. This will probably mean you need to update some application level code in order to actually use it.
