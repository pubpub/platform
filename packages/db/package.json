{
	"name": "db",
	"private": true,
	"version": "0.0.0",
	"description": "",
	"files": [
		"dist",
		"Database",
		"public",
		"table-names",
		"types"
	],
	"scripts": {
		"format": "prettier --check . --ignore-path ../../.gitignore",
		"format:fix": "prettier -w . --ignore-path ../../.gitignore",
		"type-check": "tsc --noEmit",
		"generate-db-table-names": "dotenv -e .env.local -e .env.development tsx ./scripts/generateDbTableNames.ts ./src/table-names.ts public | pino-pretty",
		"make-kysely-types": "dotenv -e .env.local -e .env.development kanel --config ./src/kanel/.kanelrc.js && pnpm generate-db-table-names && prettier -w src/"
	},
	"author": "Knowledge Futures, Inc <hello@knowledgefutures.org>",
	"license": "GPL-2.0+",
	"dependencies": {
		"logger": "workspace:*",
		"utils": "workspace:*"
	},
	"peerDependencies": {
		"@ts-rest/core": "catalog:",
		"kysely": "^0.27.2",
		"zod": "catalog:"
	},
	"devDependencies": {
		"@pubpub/prettier-config": "workspace:*",
		"@ts-rest/core": "catalog:",
		"@types/pg": "^8.11.6",
		"dotenv-cli": "^7.2.1",
		"eslint": "catalog:",
		"kanel": "^3.8.2",
		"kanel-kysely": "^0.4.0",
		"kanel-zod": "^1.3.4",
		"pg": "^8.11.3",
		"prettier": "catalog:",
		"tsconfig": "workspace:*",
		"tsx": "catalog:",
		"typescript": "catalog:",
		"zod": "catalog:"
	},
	"prettier": "@pubpub/prettier-config",
	"preconstruct": {
		"entrypoints": [
			"./Database.ts",
			"./public.ts",
			"./table-names.ts",
			"./types/index.ts"
		]
	}
}
