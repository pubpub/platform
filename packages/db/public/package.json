{
	"main": "dist/db-public.cjs.js",
	"module": "dist/db-public.esm.js"
}
