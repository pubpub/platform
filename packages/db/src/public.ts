export * from "./public/Action";
export * from "./public/ActionInstances";
export * from "./public/ActionRuns";
export * from "./public/ActionRunStatus";
export * from "./public/ApiAccessLogs";
export * from "./public/ApiAccessPermissions";
export * from "./public/ApiAccessScope";
export * from "./public/ApiAccessTokens";
export * from "./public/ApiAccessType";
export * from "./public/AuthTokens";
export * from "./public/AuthTokenType";
export * from "./public/Communities";
export * from "./public/CommunityMemberships";
export * from "./public/CoreSchemaType";
export * from "./public/ElementType";
export * from "./public/Event";
export * from "./public/FormAccessType";
export * from "./public/FormElements";
export * from "./public/FormMemberships";
export * from "./public/Forms";
export * from "./public/InputComponent";
export * from "./public/IntegrationInstances";
export * from "./public/IntegrationInstanceState";
export * from "./public/IntegrationInstanceToPub";
export * from "./public/Integrations";
export * from "./public/MemberGroups";
export * from "./public/MemberGroupToUser";
export * from "./public/MemberRole";
export * from "./public/MembershipCapabilities";
export * from "./public/MoveConstraint";
export * from "./public/PrismaMigrations";
export * from "./public/PubFields";
export * from "./public/PubFieldSchema";
export * from "./public/PubFieldToPubType";
export * from "./public/PublicSchema";
export * from "./public/PubMemberships";
export * from "./public/Pubs";
export * from "./public/PubsInStages";
export * from "./public/PubTypes";
export * from "./public/PubValues";
export * from "./public/Rules";
export * from "./public/Sessions";
export * from "./public/StageMemberships";
export * from "./public/Stages";
export * from "./public/StructuralFormElement";
export * from "./public/Users";
