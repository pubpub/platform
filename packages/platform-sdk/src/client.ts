import { initClient } from "@ts-rest/core";

import type { CommunitySpecificTypes, NonGenericProcessedPub } from "contracts";
import { createSiteApi } from "contracts";

export const makeClient = <T extends CommunitySpecificTypes = {}>(baseUrl: string) => {
	const api = createSiteApi<T>();
	// do some other modifications, such as replacing the communitySlug with the actual community slug
	// so you don't constantly have to pass it in

	return initClient(api, {
		baseUrl: baseUrl,
	});
};

export const isPubOfType = <
	T extends {
		pubType: {
			name: string;
		};
		values: any[];
	},
	PubTypeName extends T["pubType"]["name"],
>(
	pub: T,
	pubType: PubTypeName
): pub is Extract<T, { pubType: { name: PubTypeName } }> => {
	return pub.pubType.name === pubType;
};
