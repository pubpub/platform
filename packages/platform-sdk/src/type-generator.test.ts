import { describe, expect, expectTypeOf, it } from "vitest";

import type { PubFields, PubTypes, PubTypesId } from "db/public";
import { CoreSchemaType } from "db/public";

import type { Pub, PubType } from "./.generated-pubtypes";
import { isPubOfType, makeClient } from "./client";
import { generateTypeDefinitions } from "./type-generator";

export const pubTypeStub = [
	{
		id: "590c7a05-d35b-49eb-8f36-af99a7705be8",
		description: null,
		name: "Basic Pub",
		communityId: "0a17cfb2-c369-4f29-8dfd-92a51bb0ee71",
		createdAt: new Date(),
		updatedAt: new Date(),
		fields: [
			{
				id: "1a2e321a-c42a-4f24-b9cf-e66368382fdc",
				name: "Title",
				slug: "test-community:title",
				schemaName: "String",
				isRelation: false,
				isTitle: true,
				schema: null,
			},
			{
				id: "c6003a3c-7fc3-488d-8220-2f467ce4ff36",
				name: "Some relation",
				slug: "test-community:some-relation",
				schemaName: "String",
				isRelation: true,
				isTitle: false,
				schema: null,
			},
		],
	},
	{
		id: "6365dfe7-9839-4d88-baab-a7d479510acd",
		description: null,
		name: "Minimal Pub",
		communityId: "0a17cfb2-c369-4f29-8dfd-92a51bb0ee71",
		createdAt: new Date(),
		updatedAt: new Date(),
		fields: [
			{
				id: "1a2e321a-c42a-4f24-b9cf-e66368382fdc",
				name: "Title",
				slug: "test-community:title",
				schemaName: "String",
				isRelation: false,
				isTitle: true,
				schema: null,
			},
		],
	},
	{
		id: "a979056f-6b85-4aa3-9a8d-78855fdad956",
		description: null,
		name: "Everything Pub",
		communityId: "0a17cfb2-c369-4f29-8dfd-92a51bb0ee71",
		createdAt: new Date(),
		updatedAt: new Date(),
		fields: [
			{
				id: "1a2e321a-c42a-4f24-b9cf-e66368382fdc",
				name: "Title",
				slug: "test-community:title",
				schemaName: "String",
				isRelation: false,
				isTitle: true,
				schema: null,
			},
			{
				id: "c6003a3c-7fc3-488d-8220-2f467ce4ff36",
				name: "Some relation",
				slug: "test-community:some-relation",
				schemaName: "String",
				isRelation: true,
				isTitle: false,
				schema: null,
			},
			{
				id: "53713c81-66f2-446e-8164-8d4d419a978b",
				name: "Ok",
				slug: "test-community:ok",
				schemaName: "Boolean",
				isRelation: false,
				isTitle: false,
				schema: null,
			},
			{
				id: "7cac2f42-dd4d-4d1f-81e9-dd34c88bdeb1",
				name: "Number",
				slug: "test-community:number",
				schemaName: "Number",
				isRelation: false,
				isTitle: false,
				schema: null,
			},
			{
				id: "a496f1d2-17c8-4525-b7cd-8d4f42529564",
				name: "Date Time",
				slug: "test-community:date-time",
				schemaName: "DateTime",
				isRelation: false,
				isTitle: false,
				schema: null,
			},
			{
				id: "6c248f98-0535-4ca8-b8f2-31e665e481e6",
				name: "Url",
				slug: "test-community:url",
				schemaName: "URL",
				isRelation: false,
				isTitle: false,
				schema: null,
			},
			{
				id: "8b81da18-021b-4737-9a91-c13e924a9667",
				name: "Email",
				slug: "test-community:email",
				schemaName: "Email",
				isRelation: false,
				isTitle: false,
				schema: null,
			},
			{
				id: "382d81f1-1c5d-4a75-b79d-2adc4784854c",
				name: "Vector3",
				slug: "test-community:vector3",
				schemaName: "Vector3",
				isRelation: false,
				isTitle: false,
				schema: null,
			},
			{
				id: "4afbd428-3c5d-4766-815c-7f5a74a34404",
				name: "Numeric Array",
				slug: "test-community:numeric-array",
				schemaName: "NumericArray",
				isRelation: false,
				isTitle: false,
				schema: null,
			},
			{
				id: "4142263b-7d5d-4652-a8e3-562a690309fe",
				name: "String Array",
				slug: "test-community:string-array",
				schemaName: "StringArray",
				isRelation: false,
				isTitle: false,
				schema: null,
			},
			{
				id: "86f1d589-ce2c-45c8-978d-2907beb495ac",
				name: "Rich Text",
				slug: "test-community:rich-text",
				schemaName: "RichText",
				isRelation: false,
				isTitle: false,
				schema: null,
			},
			{
				id: "47c32939-02f3-4e32-9b5b-51484c20d82b",
				name: "Member",
				slug: "test-community:member",
				schemaName: "MemberId",
				isRelation: false,
				isTitle: false,
				schema: null,
			},
			{
				id: "fc498e7a-ef06-463e-aed5-83a329fb0b07",
				name: "Fun Relation",
				slug: "test-community:fun-relation",
				schemaName: "Null",
				isRelation: true,
				isTitle: false,
				schema: null,
			},
		],
	},
] as unknown as (PubTypes & { fields: (PubFields & { isTitle: true })[] })[];

describe("exportPubTypes", () => {
	it("should be able to export pub types", async () => {
		const path = new URL(".generated-pubtypes.ts", import.meta.url).pathname;

		const pubTypes = await generateTypeDefinitions(pubTypeStub, path);

		expect(pubTypes).toMatchSnapshot();
	});

	it("hngg", async () => {
		let pub = {} as unknown as Pub;

		if (pub.pubTypeName === "Everything Pub") {
			pub.values;
		}
	});
});

declare const client: ReturnType<typeof makeClient<{ Pub: Pub; PubType: PubType }>>;

const pubTypeNames = ["Basic Pub", "Minimal Pub", "Everything Pub"] as const;

describe("type tests", () => {
	describe("pub", () => {
		it("test pub Get return ", async () => {
			type PubGetReturn = Awaited<ReturnType<typeof client.pubs.get>>;

			const pub = {} as unknown as PubGetReturn;

			if (pub.status !== 200) {
				// for get body is only defined on 200
				expectTypeOf(pub.body).toEqualTypeOf<unknown>();
				return;
			}

			expectTypeOf(pub.body.pubTypeName).toMatchTypeOf<(typeof pubTypeNames)[number]>();

			if (pub.body.pubType.name === "Basic Pub") {
				pub.body.values.map((value) => {
					// this type narrowing shouldn't work
					expectTypeOf(value.fieldName).not.toEqualTypeOf<"Title" | "Some relation">();
				});
			}

			// you need to use the type guard to get the type narrowing
			if (isPubOfType(pub.body, "Basic Pub")) {
				pub.body.values.map((value) => {
					expectTypeOf(value.fieldName).toEqualTypeOf<"Title" | "Some relation">();
				});
			} else if (isPubOfType(pub.body, "Everything Pub")) {
				pub.body.values.map((value) => {
					expectTypeOf(value.fieldName).toEqualTypeOf<
						| "Title"
						| "Some relation"
						| "Ok"
						| "Number"
						| "Date Time"
						| "Url"
						| "Email"
						| "Vector3"
						| "Numeric Array"
						| "String Array"
						| "Rich Text"
						| "Member"
						| "Fun Relation"
					>();

					// this tests whether you can get the type of a field by their fieldname
					switch (value.fieldName) {
						case "Title":
							expectTypeOf(value.value).toEqualTypeOf<string>();
							// non relation fields should not have a relatedPub or relatedPubId
							expectTypeOf(value.relatedPub).toEqualTypeOf<never | undefined>();
							expectTypeOf(value.relatedPubId).toEqualTypeOf<null>();
							break;
						case "Some relation":
							// relations should have a relatedPubId and possibly a relatedPub
							expectTypeOf(value.relatedPubId).toEqualTypeOf<PubTypesId | null>();
							expectTypeOf(value.relatedPub).toEqualTypeOf<Pub | null>();
							break;
						case "Ok":
							expectTypeOf(value.value).toEqualTypeOf<boolean>();
							break;
						case "Number":
							expectTypeOf(value.value).toEqualTypeOf<number>();
							break;
						case "Date Time":
							// hmm, maybe this should be a string, unless we hydrate values on the client
							expectTypeOf(value.value).toEqualTypeOf<Date>();
							break;
						case "Url":
							expectTypeOf(value.value).toEqualTypeOf<string>();
							break;
						case "Email":
							expectTypeOf(value.value).toEqualTypeOf<string>();
							break;

						case "Vector3":
							expectTypeOf(value.value).toEqualTypeOf<[number, number, number]>();
							break;
						case "Numeric Array":
							expectTypeOf(value.value).toEqualTypeOf<number[]>();
							break;
						case "String Array":
							expectTypeOf(value.value).toEqualTypeOf<string[]>();
							break;
						case "Rich Text":
							expectTypeOf(value.value).toEqualTypeOf<{
								content: unknown[];
								type: "doc";
							}>();
							break;
						case "Member":
							expectTypeOf(value.value).toEqualTypeOf<string>();
							break;
						case "Fun Relation":
							expectTypeOf(value.value).toEqualTypeOf<null>();
							expectTypeOf(value.relatedPubId).toEqualTypeOf<PubTypesId | null>();
							expectTypeOf(value.relatedPub).toEqualTypeOf<Pub | null>();
							break;

						default:
							const _exhaustiveCheck: never = value;
							break;
					}

					// this checks whether you can get the type of a field by its CoreSchematype
					switch (value.schemaName) {
						case CoreSchemaType.Boolean:
							expectTypeOf(value.value).toEqualTypeOf<boolean>();
							expectTypeOf(value.fieldName).toEqualTypeOf<"Ok">();
							break;
						case CoreSchemaType.String:
							expectTypeOf(value.value).toEqualTypeOf<string>();
							expectTypeOf(value.fieldName).toEqualTypeOf<
								"Title" | "Some relation"
							>();
							break;
						// don't feel like testing everything here, it's mostly the same as above
						default:
							break;
					}

					switch (value.fieldSlug) {
						case "test-community:numeric-array":
							expectTypeOf(value.value).toEqualTypeOf<number[]>();
							expectTypeOf(value.fieldName).toEqualTypeOf<"Numeric Array">();
							break;
						default:
							break;
					}
				});
			}
		});
	});

	describe("pubType", () => {
		it("test pubType Get return ", async () => {
			type PubTypeGetReturn = Awaited<ReturnType<typeof client.pubTypes.get>>;

			const pubType = {} as unknown as PubTypeGetReturn;

			if (pubType.status !== 200) {
				expectTypeOf(pubType.body).toEqualTypeOf<unknown>();
				return;
			}

			expectTypeOf(pubType.body.name).toMatchTypeOf<(typeof pubTypeNames)[number]>();

			if (pubType.body.name === "Basic Pub") {
				pubType.body.fields.map((field) => {
					expectTypeOf(field.name).toEqualTypeOf<"Title" | "Some relation">();
				});
			}
		});
	});
});
