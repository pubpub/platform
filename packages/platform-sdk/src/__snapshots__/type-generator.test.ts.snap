// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`exportPubTypes > should be able to export pub types 1`] = `
"// Generated by PubPub Type Generator
import type { CoreSchemaType, PubFieldsId, PubTypesId, CommunitiesId, StagesId, PubsId } from "db/public";

export type Prettify<T> = {
    [K in keyof T]: T[K];
} & {};

export interface PubTypeField {
    id: PubFieldsId;
    name: string;
    slug: string;
    schemaName: CoreSchemaType;
    isRelation: boolean;
    isTitle: boolean;
}

export interface PubTypeDefinition {
    createdAt: Date;
    updatedAt: Date;
    id: PubTypesId;
    communityId: CommunitiesId;
    name: string;
    description: string | null;
    fields: PubTypeField[];
}

export type SchemaNameToTypeMap = {
    [CoreSchemaType.Boolean]: boolean;
    [CoreSchemaType.DateTime]: Date;
    [CoreSchemaType.Email]: string;
    [CoreSchemaType.MemberId]: string;
    [CoreSchemaType.Number]: number;
    [CoreSchemaType.NumericArray]: number[];
    [CoreSchemaType.RichText]: { type: "doc"; content: unknown[] };
    [CoreSchemaType.String]: string;
    [CoreSchemaType.StringArray]: string[];
    [CoreSchemaType.Vector3]: [number, number, number];
    [CoreSchemaType.URL]: string;
    [CoreSchemaType.Null]: null;
    [CoreSchemaType.FileUpload]: {
        id: string;
        name: string;
        type: string;
        size: number;
        url: string;
    };
}

export type PubField_title = {
    id: string;
    name: "Title";
    slug: "test-community:title";
    schemaName: CoreSchemaType.String;
    isRelation: false;
    isTitle: true;
}

export type PubField_some_relation = {
    id: string;
    name: "Some relation";
    slug: "test-community:some-relation";
    schemaName: CoreSchemaType.String;
    isRelation: true;
    isTitle: false;
}

export type PubField_ok = {
    id: string;
    name: "Ok";
    slug: "test-community:ok";
    schemaName: CoreSchemaType.Boolean;
    isRelation: false;
    isTitle: false;
}

export type PubField_number = {
    id: string;
    name: "Number";
    slug: "test-community:number";
    schemaName: CoreSchemaType.Number;
    isRelation: false;
    isTitle: false;
}

export type PubField_date_time = {
    id: string;
    name: "Date Time";
    slug: "test-community:date-time";
    schemaName: CoreSchemaType.DateTime;
    isRelation: false;
    isTitle: false;
}

export type PubField_url = {
    id: string;
    name: "Url";
    slug: "test-community:url";
    schemaName: CoreSchemaType.URL;
    isRelation: false;
    isTitle: false;
}

export type PubField_email = {
    id: string;
    name: "Email";
    slug: "test-community:email";
    schemaName: CoreSchemaType.Email;
    isRelation: false;
    isTitle: false;
}

export type PubField_vector3 = {
    id: string;
    name: "Vector3";
    slug: "test-community:vector3";
    schemaName: CoreSchemaType.Vector3;
    isRelation: false;
    isTitle: false;
}

export type PubField_numeric_array = {
    id: string;
    name: "Numeric Array";
    slug: "test-community:numeric-array";
    schemaName: CoreSchemaType.NumericArray;
    isRelation: false;
    isTitle: false;
}

export type PubField_string_array = {
    id: string;
    name: "String Array";
    slug: "test-community:string-array";
    schemaName: CoreSchemaType.StringArray;
    isRelation: false;
    isTitle: false;
}

export type PubField_rich_text = {
    id: string;
    name: "Rich Text";
    slug: "test-community:rich-text";
    schemaName: CoreSchemaType.RichText;
    isRelation: false;
    isTitle: false;
}

export type PubField_member = {
    id: string;
    name: "Member";
    slug: "test-community:member";
    schemaName: CoreSchemaType.MemberId;
    isRelation: false;
    isTitle: false;
}

export type PubField_fun_relation = {
    id: string;
    name: "Fun Relation";
    slug: "test-community:fun-relation";
    schemaName: CoreSchemaType.Null;
    isRelation: true;
    isTitle: false;
};

export type PubField = PubField_title | PubField_some_relation | PubField_ok | PubField_number | PubField_date_time | PubField_url | PubField_email | PubField_vector3 | PubField_numeric_array | PubField_string_array | PubField_rich_text | PubField_member | PubField_fun_relation;

export type PubType_Basic_Pub = {
            id: PubTypesId;
            createdAt: Date;
            updatedAt: Date;
            communityId: CommunitiesId;
            name: "Basic Pub";
            description: null;
            fields: (PubField_title | PubField_some_relation)[];
        }

export type PubType_Minimal_Pub = {
            id: PubTypesId;
            createdAt: Date;
            updatedAt: Date;
            communityId: CommunitiesId;
            name: "Minimal Pub";
            description: null;
            fields: (PubField_title)[];
        }

export type PubType_Everything_Pub = {
            id: PubTypesId;
            createdAt: Date;
            updatedAt: Date;
            communityId: CommunitiesId;
            name: "Everything Pub";
            description: null;
            fields: (PubField_title | PubField_some_relation | PubField_ok | PubField_number | PubField_date_time | PubField_url | PubField_email | PubField_vector3 | PubField_numeric_array | PubField_string_array | PubField_rich_text | PubField_member | PubField_fun_relation)[];
        };

export type PubType = PubType_Basic_Pub | PubType_Minimal_Pub | PubType_Everything_Pub

export type PubFieldToValueType<T extends PubField> = Prettify<{
    id: PubFieldsId;
    createdAt: Date;
    updatedAt: Date;
    fieldSlug: T["slug"];
    fieldName: T["name"];
    schemaName: T["schemaName"];
    value: SchemaNameToTypeMap[T["schemaName"]];
} & (T["isRelation"] extends true ? {
    relatedPubId: PubTypesId | null;
	relatedPub: Pub | null;
    } : {
      relatedPubId: null;
	  relatedPub?: never;
})>

export type Pub_Basic_Pub = {
    id: PubsId;
	title: string;
	parentId: PubsId;
	depth: number;
    pubTypeId: PubTypesId;
    stageId: StagesId;
    communityId: CommunitiesId;
    createdAt: Date;
    updatedAt: Date;
    pubTypeName: "Basic Pub";
    pubType: PubType_Basic_Pub;
    values: (PubFieldToValueType<PubField_title> | PubFieldToValueType<PubField_some_relation>)[]
}

export type Pub_Minimal_Pub = {
    id: PubsId;
	title: string;
	parentId: PubsId;
	depth: number;
    pubTypeId: PubTypesId;
    stageId: StagesId;
    communityId: CommunitiesId;
    createdAt: Date;
    updatedAt: Date;
    pubTypeName: "Minimal Pub";
    pubType: PubType_Minimal_Pub;
    values: (PubFieldToValueType<PubField_title>)[]
}

export type Pub_Everything_Pub = {
    id: PubsId;
	title: string;
	parentId: PubsId;
	depth: number;
    pubTypeId: PubTypesId;
    stageId: StagesId;
    communityId: CommunitiesId;
    createdAt: Date;
    updatedAt: Date;
    pubTypeName: "Everything Pub";
    pubType: PubType_Everything_Pub;
    values: (PubFieldToValueType<PubField_title> | PubFieldToValueType<PubField_some_relation> | PubFieldToValueType<PubField_ok> | PubFieldToValueType<PubField_number> | PubFieldToValueType<PubField_date_time> | PubFieldToValueType<PubField_url> | PubFieldToValueType<PubField_email> | PubFieldToValueType<PubField_vector3> | PubFieldToValueType<PubField_numeric_array> | PubFieldToValueType<PubField_string_array> | PubFieldToValueType<PubField_rich_text> | PubFieldToValueType<PubField_member> | PubFieldToValueType<PubField_fun_relation>)[]
}

export type Pub = Pub_Basic_Pub | Pub_Minimal_Pub | Pub_Everything_Pub


export interface CommunitySchema {
    pubTypes: {
        "Basic Pub": {
                id: "590c7a05-d35b-49eb-8f36-af99a7705be8",
                name: "Basic Pub",
                description: null,
                fields: {
                    "test-community:title": {
                        id: "1a2e321a-c42a-4f24-b9cf-e66368382fdc",
                        name: "Title",
                        slug: "test-community:title",
                        schemaName: CoreSchemaType.String,
                        isRelation: false,
                        isTitle: true
                    },
                    "test-community:some-relation": {
                        id: "c6003a3c-7fc3-488d-8220-2f467ce4ff36",
                        name: "Some relation",
                        slug: "test-community:some-relation",
                        schemaName: CoreSchemaType.String,
                        isRelation: true,
                        isTitle: false
                    }
                }
            },
        "Minimal Pub": {
                id: "6365dfe7-9839-4d88-baab-a7d479510acd",
                name: "Minimal Pub",
                description: null,
                fields: {
                    "test-community:title": {
                        id: "1a2e321a-c42a-4f24-b9cf-e66368382fdc",
                        name: "Title",
                        slug: "test-community:title",
                        schemaName: CoreSchemaType.String,
                        isRelation: false,
                        isTitle: true
                    }
                }
            },
        "Everything Pub": {
                id: "a979056f-6b85-4aa3-9a8d-78855fdad956",
                name: "Everything Pub",
                description: null,
                fields: {
                    "test-community:title": {
                        id: "1a2e321a-c42a-4f24-b9cf-e66368382fdc",
                        name: "Title",
                        slug: "test-community:title",
                        schemaName: CoreSchemaType.String,
                        isRelation: false,
                        isTitle: true
                    },
                    "test-community:some-relation": {
                        id: "c6003a3c-7fc3-488d-8220-2f467ce4ff36",
                        name: "Some relation",
                        slug: "test-community:some-relation",
                        schemaName: CoreSchemaType.String,
                        isRelation: true,
                        isTitle: false
                    },
                    "test-community:ok": {
                        id: "53713c81-66f2-446e-8164-8d4d419a978b",
                        name: "Ok",
                        slug: "test-community:ok",
                        schemaName: CoreSchemaType.Boolean,
                        isRelation: false,
                        isTitle: false
                    },
                    "test-community:number": {
                        id: "7cac2f42-dd4d-4d1f-81e9-dd34c88bdeb1",
                        name: "Number",
                        slug: "test-community:number",
                        schemaName: CoreSchemaType.Number,
                        isRelation: false,
                        isTitle: false
                    },
                    "test-community:date-time": {
                        id: "a496f1d2-17c8-4525-b7cd-8d4f42529564",
                        name: "Date Time",
                        slug: "test-community:date-time",
                        schemaName: CoreSchemaType.DateTime,
                        isRelation: false,
                        isTitle: false
                    },
                    "test-community:url": {
                        id: "6c248f98-0535-4ca8-b8f2-31e665e481e6",
                        name: "Url",
                        slug: "test-community:url",
                        schemaName: CoreSchemaType.URL,
                        isRelation: false,
                        isTitle: false
                    },
                    "test-community:email": {
                        id: "8b81da18-021b-4737-9a91-c13e924a9667",
                        name: "Email",
                        slug: "test-community:email",
                        schemaName: CoreSchemaType.Email,
                        isRelation: false,
                        isTitle: false
                    },
                    "test-community:vector3": {
                        id: "382d81f1-1c5d-4a75-b79d-2adc4784854c",
                        name: "Vector3",
                        slug: "test-community:vector3",
                        schemaName: CoreSchemaType.Vector3,
                        isRelation: false,
                        isTitle: false
                    },
                    "test-community:numeric-array": {
                        id: "4afbd428-3c5d-4766-815c-7f5a74a34404",
                        name: "Numeric Array",
                        slug: "test-community:numeric-array",
                        schemaName: CoreSchemaType.NumericArray,
                        isRelation: false,
                        isTitle: false
                    },
                    "test-community:string-array": {
                        id: "4142263b-7d5d-4652-a8e3-562a690309fe",
                        name: "String Array",
                        slug: "test-community:string-array",
                        schemaName: CoreSchemaType.StringArray,
                        isRelation: false,
                        isTitle: false
                    },
                    "test-community:rich-text": {
                        id: "86f1d589-ce2c-45c8-978d-2907beb495ac",
                        name: "Rich Text",
                        slug: "test-community:rich-text",
                        schemaName: CoreSchemaType.RichText,
                        isRelation: false,
                        isTitle: false
                    },
                    "test-community:member": {
                        id: "47c32939-02f3-4e32-9b5b-51484c20d82b",
                        name: "Member",
                        slug: "test-community:member",
                        schemaName: CoreSchemaType.MemberId,
                        isRelation: false,
                        isTitle: false
                    },
                    "test-community:fun-relation": {
                        id: "fc498e7a-ef06-463e-aed5-83a329fb0b07",
                        name: "Fun Relation",
                        slug: "test-community:fun-relation",
                        schemaName: CoreSchemaType.Null,
                        isRelation: true,
                        isTitle: false
                    }
                }
            }
    }
}
"
`;
